src/components/ColorPicker/index.tsx

```diff
@@ -0,0 +1,185 @@
+import React, { useEffect, useRef, useState } from "react";
+import { ActionCreator } from "redux";
+import { Popover, Tooltip } from "antd";
+import { HexColorInput, HexColorPicker } from "react-colorful";
+import classNames from "classnames";
+import { useDebounce } from "use-debounce";
+import { ColorChange } from "@aics/simularium-viewer";
+
+import { AGENT_COLORS } from "../../containers/ViewerPanel/constants";
+import {
+    SetColorChangeAction,
+    SetRecentColorsAction,
+} from "../../state/selection/types";
+
+import styles from "./style.css";
+
+interface ColorPickerProps {
+    childrenHaveDifferentColors?: boolean;
+    selectedColor: string;
+    agentName: string;
+    tags: string[];
+    recentColors: string[];
+    setColorChange: ActionCreator<SetColorChangeAction>;
+    setRecentColors: ActionCreator<SetRecentColorsAction>;
+}
+
+const ColorPicker = ({
+    agentName,
+    tags,
+    recentColors,
+    setColorChange,
+    setRecentColors,
+    selectedColor: initialColor,
+    childrenHaveDifferentColors,
+}: ColorPickerProps) => {
+    const [currentColor, setCurrentColor] = useState(initialColor);
+    const [debouncedColor] = useDebounce(currentColor, 250);
+    const isInitialRender = useRef(true);
+    const [isColorPickerVisible, setColorPickerVisible] = useState(false);
+    const [lastSelectedColor, setLastSelectedColor] = useState(initialColor);
+
+    const togglePopover = () => {
+        if (isColorPickerVisible) {
+            setColorPickerVisible(false);
+        } else {
+            setLastSelectedColor(currentColor);
+            setColorPickerVisible(true);
+        }
+    };
+
+    const handleColorChange = (newColor: string) => {
+        const colorChange: ColorChange = {
+            agent: { name: agentName, tags: tags },
+            color: newColor.toLowerCase(),
+        };
+        setColorChange(colorChange);
+    };
+
+    useEffect(() => {
+        if (isInitialRender.current) {
+            isInitialRender.current = false;
+        } else {
+            handleColorChange(debouncedColor);
+            updateRecentColors(debouncedColor);
+        }
+    }, [debouncedColor]);
+
+    useEffect(() => {
+        setCurrentColor(initialColor);
+    }, [initialColor]);
+
+    const updateRecentColors = (color: string) => {
+        if (recentColors.includes(color)) {
+            return;
+        }
+        const newRecentColors = [color, ...recentColors];
+        if (newRecentColors.length > 18) {
+            newRecentColors.pop();
+        }
```

Comment:
nit: even though something truly terrible would have to happen to make it run more than once, I'd still use `while` here for extra safety.